### PR TITLE
Get refcode on MacOS from xattr.

### PR DESCRIPTION
**Pull request by mkarolin**
Resolves https://github.com/brave/brave-browser/issues/32316

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: base/mac/files/file_extended_attribute_unittest.cc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+#include "base/files/file_path.h"
+#include "base/files/file_util.h"
+#include "base/files/scoped_temp_dir.h"
+#include "brave/base/mac/files/file_extended_attribute.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace {
+const char kXattrName[] = "com.brave.refcode";
+const char kXattrValue[] = "0xDEADFACE";
```
Comment (simonhong) [inline base/mac/files/file_extended_attribute_unittest.cc:18]: nit: `constexpr`

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 2]
```diff
@@ -125,9 +132,47 @@ void DeletePromoCodeFile(const base::FilePath& promo_code_file) {
   }
 }
 
+#if BUILDFLAG(IS_MAC)
+std::string ReadPromoCodeFromXattr() {
+  static const char kRefCodeAttr[] = "com.brave.refcode";
```
Comment (simonhong) [inline components/brave_referrals/browser/brave_referrals_service.cc:137]: nit:`constexpr`

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 3]
```diff
@@ -125,9 +132,47 @@ void DeletePromoCodeFile(const base::FilePath& promo_code_file) {
...
+    if (promo_code.empty()) {
+      LOG(ERROR) << "Promo code value from " << bundle_path
+                 << "'s extended attribute " << kRefCodeAttr << " is empty.";
+    } else {
+      DVLOG(1) << "Promo code from " << kRefCodeAttr << ": " << promo_code;
+      return promo_code;
+    }
+  }
+
+  return "";
```
Comment (simonhong) [inline components/brave_referrals/browser/brave_referrals_service.cc:162]: nit: `base::EmptyString();`

**Review by simonhong — APPROVED** [review]
++

File: components/brave_referrals/browser/brave_referrals_service.cc (hunk 2 above)
Comment (mkarolin) [reply components/brave_referrals/browser/brave_referrals_service.cc]: Fixed in [2ac2c5f](https://github.com/brave/brave-core/commit/2ac2c5f7d14ac1ec18a13631bc910e6276fc04e6)

File: components/brave_referrals/browser/brave_referrals_service.cc (hunk 3 above)
Comment (mkarolin) [reply components/brave_referrals/browser/brave_referrals_service.cc:162]: From the upstream comment it seems that we shouldn't use `base::EmptyString()` in this case (https://source.chromium.org/chromium/chromium/src/+/main:base/strings/string_util.h;l=204): 

> ...There is only one case where you should use these: functions
// which need to return a string by reference (e.g. as a class member
// accessor), and don't have an empty string to use (e.g. in an error case).
// These should not be used as initializers, function arguments, or return
// values for functions which return by value or outparam.

Comment (simonhong) [reply components/brave_referrals/browser/brave_referrals_service.cc:162]: Got it! Thanks for checking 👍🏼

File: base/mac/files/file_extended_attribute.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+namespace brave {
+
+// Extracts the value of an extended file attribute specified by |name| from the
+// file specified by the |path|. The result is placed in the |value|. On
+// success returns 0. Otherwise, returns ERRNO value. Error values of
+// interest:
+// ENOATTR - the attribute with the given |name| was not found.
+// ENOTSUP - the file system doesn't support (or disabled) extended attributes.
+// ENODATA - the value could not be retrieved.
+int GetFileExtendedAttribute(const base::FilePath& path,
```
Comment (bridiver) [inline base/mac/files/file_extended_attribute.h:22]: I don't really see any reason to expose this stuff in base, I would just keep it all in the referrals service

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 5]
```diff
@@ -125,9 +132,47 @@ void DeletePromoCodeFile(const base::FilePath& promo_code_file) {
...
 
+#if BUILDFLAG(IS_MAC)
+std::string ReadPromoCodeFromXattr() {
+  static constexpr char kRefCodeAttr[] = "com.brave.refcode";
+  base::FilePath bundle_path = base::apple::OuterBundlePath();
+  std::vector<char> value;
+  int result_errno =
+      brave::GetFileExtendedAttribute(bundle_path, kRefCodeAttr, &value);
+  if (result_errno == ENOATTR) {
+    LOG(WARNING) << "Could not get promo code from " << bundle_path
```
Comment (bridiver) [inline components/brave_referrals/browser/brave_referrals_service.cc:143]: VLOG

File: base/mac/files/file_extended_attribute.h (hunk 4 above)
Comment (mkarolin) [reply base/mac/files/file_extended_attribute.h:22]: It seemed like a more generic functionality that might be useful elsewhere at some point. Also, makes it more awkward to unit test - I'd have to expose it in the referrals service header.

File: components/brave_referrals/browser/brave_referrals_service.cc (hunk 5 above)
Comment (mkarolin) [reply components/brave_referrals/browser/brave_referrals_service.cc]: Fixed in https://github.com/brave/brave-core/pull/20262/commits/83030888a121b765276ee249336f92af0561e05f

File: base/mac/files/file_extended_attribute.h (hunk 4 above)
Comment (mkarolin) [reply base/mac/files/file_extended_attribute.h]: Moved these files under `components/brave_referrals/browser/` in https://github.com/brave/brave-core/pull/20262/commits/83030888a121b765276ee249336f92af0561e05f

File: components/brave_referrals/browser/brave_referrals_service.cc [hunk 6]
```diff
@@ -125,9 +132,46 @@ void DeletePromoCodeFile(const base::FilePath& promo_code_file) {
...
+  static constexpr char kRefCodeAttr[] = "com.brave.refcode";
+  base::FilePath bundle_path = base::apple::OuterBundlePath();
+  std::vector<char> value;
+  int result_errno =
+      brave::GetFileExtendedAttribute(bundle_path, kRefCodeAttr, &value);
+  if (result_errno == ENOATTR) {
+    VLOG(0) << "Could not get promo code from " << bundle_path
+            << ". The extended attribute " << kRefCodeAttr << " was not found.";
+  } else if (result_errno != 0) {
+    LOG(ERROR) << "Could not get promo code from " << bundle_path
```
Comment (bridiver) [inline components/brave_referrals/browser/brave_referrals_service.cc:146]: sorry, I mean all of these should be VLOG

File: components/brave_referrals/browser/file_extended_attribute.cc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  if (expected_length < 0) {
+    return errno;
+  }
+
+  value->resize(expected_length);
+  ssize_t length = getxattr(path.value().c_str(), name, value->data(),
+                            expected_length, /*position=*/0,
+                            /*options=*/0);
+  if (length != expected_length) {
+    LOG(ERROR) << "Failed to retrive extended attribute " << name
```
Comment (bridiver) [inline components/brave_referrals/browser/file_extended_attribute.cc:30]: VLOG

**Review by bridiver — APPROVED** [review]
fix LOG -> VLOG and then lgtm

File: components/brave_referrals/browser/brave_referrals_service.cc (hunk 6 above)
Comment (mkarolin) [reply components/brave_referrals/browser/brave_referrals_service.cc]: Fixed in [18fa76b](https://github.com/brave/brave-core/commit/18fa76b2fafc403c44d02bc93875a70814f4c40f)

File: components/brave_referrals/browser/file_extended_attribute.cc (hunk 7 above)
Comment (mkarolin) [reply components/brave_referrals/browser/file_extended_attribute.cc]: Fixed in [18fa76b](https://github.com/brave/brave-core/commit/18fa76b2fafc403c44d02bc93875a70814f4c40f)

**Pull request merged**